cirq/study/flattener.py

```diff
@@ -0,0 +1,275 @@
+
+from cirq.study.resolver import ParamResolver, ParamResolverOrSimilarType
+from cirq.study.sweeps import Sweep, Params
+from cirq.study.sweepable import to_sweep
+from cirq.protocols.resolve_parameters import resolve_parameters
+
+
+class ParamFlattener(ParamResolver):
+    """A `ParamResolver` that resolves sympy expressions to unique symbols.
+
+    This is a mutable object that stores new expression to symbol mappings
+    when it is used to resolve parameters with `cirq.resolve_parameters` or
+    `ParamFlattener.flatten_circuit`.  It is useful for replacing sympy
+    expressions from circuits with single symbols and transforming parameter
+    sweeps to match.
+
+    Example:
+        >>> qubit = cirq.LineQubit(0)
+        >>> a = sympy.Symbol('a')
+        >>> circuit = cirq.Circuit.from_ops(
+        ...     cirq.X(qubit) ** (a/4),
+        ...     cirq.X(qubit) ** (1-a/4),
+        ... )
+        >>> print(circuit)
+        0: ───X^(a/4)───X^(1 - a/4)───
+
+        >>> sweep = cirq.Linspace(a, start=0, stop=3, length=4)
+        >>> list(sweep)  #doctest: +NORMALIZE_WHITESPACE
+        [cirq.ParamResolver(OrderedDict([('a', 0.0)])),
+         cirq.ParamResolver(OrderedDict([('a', 1.0)])),
+         cirq.ParamResolver(OrderedDict([('a', 2.0)])),
+         cirq.ParamResolver(OrderedDict([('a', 3.0)]))]
+
+        >>> flattener = cirq.ParamFlattener()
+        >>> c_flat = flattener.flatten(circuit)
+        >>> print(c_flat)
+        0: ───X^x0───X^x1───
+
+        >>> new_sweep = flattener.transform_sweep(sweep)
+        >>> list(new_sweep)  #doctest: +NORMALIZE_WHITESPACE
+        [cirq.ParamResolver(OrderedDict([('x0', 0.0), ('x1', 1.0)])),
+         cirq.ParamResolver(OrderedDict([('x0', 0.25), ('x1', 0.75)])),
+         cirq.ParamResolver(OrderedDict([('x0', 0.5), ('x1', 0.5)])),
+         cirq.ParamResolver(OrderedDict([('x0', 0.75), ('x1', 0.25)]))]
+
+        >>> print(cirq.resolve_parameters(c_flat, list(new_sweep)[1]))
+        0: ───X^0.25───X^0.75───
+        >>> print(cirq.resolve_parameters(circuit, list(sweep)[1])) # Equivalent
+        0: ───X^0.25───X^0.75───
+    """
+
+    def __new__(cls, *args, **kwargs):
+        """Disables the behavior of `ParamResolver.__new__`."""
+        return super().__new__(cls)
+
+    def __init__(self,
+                 param_dict: Optional[ParamResolverOrSimilarType] = None,
+                 new_param_names: Iterable[str] = ()):
+        """Initializes a new ParamFlattener.
+
+        Args:
+            param_dict: A default initial mapping from some parameter names,
+                symbols, or expressions to other symbols or values.  Only sympy
+                expressions and symbols not specified in `param_dict` will be
+                flattened.
+            new_param_names: An iterator of parameter names to replace each
+                expression with.  It may be infinite but there must be no
+                repetitions.  By default the parameter names are x0, x1, x2, ...
```

Comment:
An alternative default would be to use the str() on the sympy formula.  That would make the mapping that this potentially more readable.

Reply:
Yes but I'm worried about symbols with names that look like expressions.  `repr(sympy.Symbol('x + 1') + 1) == 'x + 1 + 1'`.  The circuit diagrams would look identical but have different meanings.

Reply:
Done.  `Used '<{}>'.format(symbol)` to avoid ambiguity.